Jenkinsfile

```diff
@@ -112,136 +101,223 @@ pipeline {
-                        when {
-                            anyOf{
-                                expression { SMOKE_TESTS_FLAG == "" };
-                                allOf {
-                                    expression { SMOKE_TESTS_FLAG == "-m smoke" };
-                                    expression { PYTHON == DEFAULT_TOX_PYTHON_VERSION }
+                        stages {
+                            stage('Set env to run only smoke tests') {
+                                when {
+                                    allOf{
+                                        not{ branch 'master' };
+                                        not{ branch 'develop' };
+                                        expression { params.TESTS == 'Smoke' }
+                                    }
+                                }
+                                steps {
+                                    script {
+                                        SMOKE_TESTS_FLAG = "-m smoke"
+                                    }
                                 }
                             }
-                        }
-                        steps {
-                            lock(getLockForProtocol(PROTOCOL)) {
-                                checkout scm
-                                script {installTox ()}
-                                bat """
-                                    venv\\Scripts\\python.exe -m tox -e py${PYTHON} -- ${SMOKE_TESTS_FLAG} --protocol ${PROTOCOL} --slave ${SLAVE} --junitxml=pytest_reports/pytest_${PROTOCOL}_${SLAVE}_report_py${PYTHON}.xml
-                                """
+                            stage('Set env to run all tests') {
+                                when {
+                                    anyOf{
+                                        branch 'master';
+                                        branch 'develop';
+                                        expression { params.TESTS == 'All' }
+                                    }
+                                }
+                                steps {
+                                    script {
+                                        SMOKE_TESTS_FLAG = ""
+                                    }
+                                }
                             }
-                        }
-                        post {
-                            always {
-                                bat """
-                                    move .coverage .coverage_${PROTOCOL}
-                                """    
-                                junit "pytest_reports\\*.xml"
-                                archiveArtifacts artifacts: 'pytest_reports\\*.xml'
-                                stash includes: ".coverage_${PROTOCOL}", name: "coverage_report_${PROTOCOL}"
+                            stage('Run tests') {
+                                options {
+                                    lock('sequential-matrix')
```

Comment:
what is this?

Reply:
This is to force that the matrix is running sequentially. I had to add it because Jenkins tried running all the steps in parallel.

Reply:
Why should they run sequentially?

Reply:
There are issues when creating the tox environments at the same time. And besides that, the tests start a connection with the real drives, so we should have one connection type of connection at a time.

Reply:
What are those issues? Matrix only runs tasks in parallel when a there are multiple nodes. If it's the same node it doesn't have any other option than to run them sequentially.

we should have one connection type of connection at a time -> Thats what the ecat_node_lock is supposed to do, isn't it?

Reply:
I guess. But it seems that the matrix is ignoring the lock somehow: See [this job](http://jenkins.ingenia/job/Novanta%20Motion%20-%20Ingenia%20-%20Git/job/ingeniamotion/job/test-INGM-338/2/console) for example.